fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexMaintainer.java

```diff
@@ -241,9 +262,28 @@ private Map<IndexOptions, List<LuceneDocumentFromRecord.DocumentField>> getIndex
     }
 
     @SuppressWarnings("PMD.CloseResource")
-    private void deleteDocument(Tuple groupingKey, byte[] primaryKey) throws IOException {
+    private void deleteDocument(Tuple groupingKey, Tuple primaryKey) throws IOException {
         final IndexWriter oldWriter = directoryManager.getIndexWriter(groupingKey, indexAnalyzerSelector.provideIndexAnalyzer(""));
-        Query query = SortedDocValuesField.newSlowExactQuery(PRIMARY_KEY_SEARCH_NAME, new BytesRef(primaryKey));
+        Query query;
+        String formatString = state.index.getOption(LuceneIndexOptions.PRIMARY_KEY_SERIALIZATION_FORMAT);
+        LuceneIndexKeySerializer ser = LuceneIndexKeySerializer.fromStringFormat(formatString, primaryKey);
+        // null format string means don't use BinaryPoint for the index primary key
+        if (formatString != null) {
+            try {
+                byte[][] binaryPoint = ser.asFormattedBinaryPoint();
+                query = BinaryPoint.newRangeQuery(PRIMARY_KEY_BINARY_POINT_NAME, binaryPoint, binaryPoint);
+            } catch (RecordCoreFormatException ex) {
+                // this can happen on format mismatch or encoding error
+                // fallback to the old way (less efficient)
+                query = SortedDocValuesField.newSlowExactQuery(PRIMARY_KEY_SEARCH_NAME, new BytesRef(ser.asPackedByteArray()));
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Failed to delete using BinaryPoint encoded ID: {}", ex.getMessage());
+                }
+            }
+        } else {
+            // fallback to the old way (less efficient)
+            query = SortedDocValuesField.newSlowExactQuery(PRIMARY_KEY_SEARCH_NAME, new BytesRef(ser.asPackedByteArray()));
+        }
         oldWriter.deleteDocuments(query);
```

Comment:
Can this be
```
        long count = oldWriter.deleteDocuments(query);
        if (count != 1) {
            throw new RecordCoreException(...);
        }
```
or is there some legit case where we call `deleteDocument` for one that does not exist in the index?

Reply:
Is this a change to the update behavior? I'd rather keep that on  another PR, if possible... This would probably need additional testing, etc.

Reply:
Yes, I do believe that an error check is appropriate here, particularly as the query logic has gotten more complicated. But I do see that there is potentially a problem with updating documents saved earlier in the same transaction. And so I am okay with taking such a check as a separate task.